vermouth/map_parser.py

```diff
@@ -0,0 +1,403 @@
+        mapped_nodes = {from_ for from_ in self.mapping if node_to in self.mapping[from_]}
+        nodes_from = nodes_from.intersection(mapped_nodes)
+        assert len(nodes_from) == 1
+        self.references[node_to] = next(iter(nodes_from))
+
+    def get_mapping(self, map_type):
+        if self.blocks_from is None:
+            return None
+        mapping = Mapping(self.blocks_from, self.blocks_to, dict(self.mapping),
+                          self.references, ff_from=self.ff_from, ff_to=self.ff_to,
+                          type=map_type, names=tuple(self.names))
+        return mapping
+
+
+class MappingDirector:
+    RESNAME_NUM_SEP = '#'
+    RESIDUE_ATOM_SEP = ':'
+    MAP_TYPES = ['block', 'modification']
+
+    def __init__(self, builder=None):
+        if builder is None:
+            self.builder = MappingBuilder()
+        else:
+            self.builder = builder
+        self.identifiers = {}
+        self.section = None
+        self.map_type = None
+        self.from_ff = None
+        self.to_ff = None
+        self.macros = {}
+        self._current_id = {'from_': None, 'to_': None}
+
+    def parse(self, file_handle):
+        for lineno, line in enumerate(file_handle, 1):
+            # TODO split off comments
+            line = line.strip()
+            if not line:
+                continue
+            if self._is_section_header(line):
+                map_type = self.map_type
+                new_mapping = self._header(line)
+                if new_mapping:
+                    self._current_id = {'from_': None, 'to_': None}
+                    self.identifiers = {}
+                    mapping = self.builder.get_mapping(map_type)
+                    if mapping is not None:
+                        yield mapping
+                        self.builder.reset()
+            else:
+                self._parse_line(line, lineno)
+        mapping = self.builder.get_mapping(self.map_type)
+        if mapping is not None:
+            yield mapping
+
+    @staticmethod
+    def _is_section_header(line):
+        return line.startswith('[') and line.endswith(']')
+    
+    @staticmethod
+    def _section_to_method(section):
+        method_name = '_' + section.replace(' ', '_')
+        return method_name
+
+    def _parse_line(self, line, lineno):
+        line = _substitute_macros(line, self.macros)
+        method_name = self._section_to_method(self.section)
+        try:
+            getattr(self, method_name)(line)
+        except Exception:
+            LOGGER.error("Problems parsing line {}. I think it should be a '{}'"
+                         " line, but I can't parse it as such.",
+                         lineno, self.section)
+            raise
+
+    def _header(self, line):
+        section = line.strip('[ ]').casefold()
+        method_name = self._section_to_method(section)
+        try:
+            if section not in self.MAP_TYPES:
+                getattr(self, method_name)
```

Comment:
This comes with side effects. I can write a mapping file with a section named `[ header ]`, `[ parse_blocks ]`, or even `[ _init__ ]` or `[ _del__ ]`. One way to solve this would be to have a method decorator that would declare a method a being the parser for a section. Something in the line of:

```python
def section_parser(method):
    name = method.__name__[1:]
    method._section_name = name
    return method


def parser_class(cls):
    if not hasattr(cls, 'METH_DICT'):
        raise TypeError('Must have a METH_DICT attribute.')
    for attribute_name in dir(cls):
        attribute = getattr(cls, attribute_name)
        if hasattr(attribute, '_section_name'):
            cls.METH_DICT[attribute._section_name] = attribute
    return cls


@parser_class
class A:
    METH_DICT = {}

    @section_parser
    def _plop(self):
        return 3
    
    @section_parser
    def _toto_plop(self, bla=42):
        return bla


print(A.METH_DICT)
```

It is not as trivial as I'd like because I wanted `METH_DICT` in the class. Though, it avoids accidental sections, and with little work on the `section_parser` decorator, it allows to decouple the name of the method and the name of the section in a per case basis. This means that we could add a `[ header ]` section without renaming the existing `_header` method.

It would also make a linter easier to write: the class then know what are the valid sections.

Reply:
Great idea.

Reply:
Something to puzzle over for the holidays: Would it make more sense to do this with a metaclass? Since the methods *should* be registered when the class is created (as opposed to instantiated). As an alternative, we should make a `Register` (or something) class with a `__call__` method that registers the methods with the correct name (subclass of dict?). Next question would be how to attach the register to the actual parser though.


For future reference: https://stackoverflow.com/a/50372503

Reply:
And, how do we extend this to make the sections `[ block ] [ mapping ]` and `[ modification ] [ mapping ]` call different functions? This'll be particularly valuable for the other parsers and file formats.
How deep can we/you nest these sections? We don't have section endings that would resolve this.